Source/ASTextNode2.mm

```diff
@@ -360,115 +366,65 @@ - (NSObject *)drawParametersForAsyncLayer:(_ASDisplayLayer *)layer
   [self _ensureTruncationText];
   ASTextContainer *copiedContainer = [_textContainer copy];
   copiedContainer.size = self.bounds.size;
-  NSMutableAttributedString *mutableText = [self.attributedText mutableCopy] ?: [[NSMutableAttributedString alloc] init];
-  
-  [self prepareAttributedString:mutableText];
   
   return @{
-    @"container": copiedContainer,
-    @"text": mutableText,
-    @"bgColor": self.backgroundColor ?: [NSNull null]
-  };
+           @"container": copiedContainer,
+           @"text": [self l_preparedAttributedText],
+           @"bgColor": self.backgroundColor ?: [NSNull null],
+           // Pass along self as a hack so that we can update _lastUsedLayout.
+           // We'll be careful
+           @"instance" : self
+           };
 }
 
 /**
  * If it can't find a compatible layout, this method creates one.
  *
- * NOTE: Be careful to copy `text` if needed.
+ * NOTE: The `container` you pass into this should be a copy, as it may be
+ * retained by the cache in the event of a cache miss.
  */
-+ (ASTextLayout *)compatibleLayoutWithContainer:(ASTextContainer *)container
+- (ASTextLayout *)compatibleLayoutWithContainer:(ASTextContainer *)container
                                            text:(NSAttributedString *)text NS_RETURNS_RETAINED
-
 {
-  // Allocate layoutCacheLock on the heap to prevent destruction at app exit (https://github.com/TextureGroup/Texture/issues/136)
-  static ASDN::StaticMutex& layoutCacheLock = *new ASDN::StaticMutex;
-  static NSCache<NSAttributedString *, ASTextCacheValue *> *textLayoutCache;
+  // First quick check – if the last layout we computed is useable.
+  // On a cache miss, this method is going to run long and in some
+  // scenarios it may make sense to run concurrently with different
+  // sizes, so don't hold the lock the whole time.
+  auto lastLayout = ASLockedSelf(_lastUsedLayout);
```

Comment:
Doesn't the caller need to have self locked this whole time? Otherwise, it will need to lock after it receives the layout, and re-check that it is still valid.

Reply:
No, because we're inside a method with immutable parameters. The validity of the layout can't change out from under us – whatever text and container they passed in, that's what we need to respond with. It's up to the consumer to worry about whether those parameters remain valid throughout the method.

I really hated to turn this from a class method to an instance method because it's so unclear that this method simply does not rely on any instance state. It asks the instance for a quick candidate layout to consult first, and it puts the result into that bucket for the next call.

For instance, the layout system calls this with the lock held for the reason you mentioned. The display system is operating on fixed draw params, so it doesn't need to lock the node to call this.